Kitodo-DataAccess/src/main/java/org/kitodo/dataaccess/storage/memory/GraphPath.java

```diff
@@ -0,0 +1,313 @@
+
+    /**
+     * This is a location step of a graph path.
+     */
+    static final NodeReference LOCATION_STEP = MemoryStorage.INSTANCE
+            .createNodeReference("http://names.kitodo.org/GraphPath/v1#LocationStep");
+
+    /**
+     * Direction of reference is forwards one step.
+     */
+    static final NodeReference TO = MemoryStorage.INSTANCE
+            .createNodeReference("http://names.kitodo.org/GraphPath/v1#to");
+
+    /**
+     * Applies a graph path on a collection of nodes.
+     *
+     * <p>
+     * Currently only supports relation type “to”.
+     *
+     * @param nodes
+     *            nodes to apply the graph path on
+     * @param path
+     *            the graph path to apply
+     * @return the matching nodes
+     */
+    public static Result apply(Set<ObjectType> nodes, Node path) {
+        MemoryResult result = new MemoryResult();
+
+        Set<Node> allTo = path.get(TO).nodes();
+
+        Result nodeResult = nodes instanceof Result ? (Result) nodes : new MemoryResult();
+
+        if (allTo.isEmpty() || nodes.isEmpty()) {
+            return nodeResult;
+        }
+
+        Set<Node> nodesOnly = nodeResult.nodes();
+
+        for (Node toSegment : allTo) {
+            MemoryResult thisResult = new MemoryResult();
+            Set<String> predicates = new HashSet<>();
+            for (IdentifiableNode predicate : toSegment.get(RDF.PREDICATE).identifiableNodes()) {
+                predicates.add(predicate.getIdentifier());
+            }
+            Collection<ObjectType> objects = new LinkedList<>();
+            toSegment.get(RDF.OBJECT).forEach(objects::add);
+            for (Node node : nodesOnly) {
+                node.get(predicates, objects).forEach(thisResult::add);
+            }
+            apply(thisResult, toSegment).forEach(result::add);
+        }
+
+        return result;
+    }
+
+    /**
+     * Replaces the prefix of the string, if any and specified in the map of
+     * prefixes.
+     *
+     * @param prefixes
+     *            map of prefixes, direction of mapping is prefix to namespace.
+     * @param string
+     *            the string to process
+     * @return the string with the prefix replaced
+     */
+    private static String applyPrefixes(Map<String, String> prefixes, String string) {
+        final int NOT_FOUND = -1;
+        int colonIndex = string.indexOf(':');
+        if (colonIndex == NOT_FOUND) {
+            return string;
+        }
+        String prefix = string.substring(0, colonIndex);
+        if (prefixes.containsKey(prefix)) {
+            return Namespaces.concat(prefixes.get(prefix), string.substring(colonIndex + 1));
+        }
+        return string;
+    }
+
+    /**
+     * Creates an empty GraphPath. An empty GraphPath points to the node it is
+     * applied on.
+     *
+     * <p>
+     * This method is used in tests to build comparable objects
+     * programmatically.
+     */
+    protected GraphPath() {
+        super(GRAPH_PATH);
```

Comment:
Why make her something what is only specifically fot tests? Can't tests use public constructor?

Reply:
no, because I need a reference object to test the object the constructor created. The reference object is created in the test class like what the constructor should deliver. For the reason that equals() is true, the comparison object must be of class GraphPath() as well.